llvm/lib/Transforms/IPO/ExpandVariadics.cpp

```diff
@@ -0,0 +1,1056 @@
+    assert(NF);
+    NewCB = InvokeInst::Create(NF, II->getNormalDest(), II->getUnwindDest(),
+                               Args, OpBundles, "", CB);
+  } else {
+    report_fatal_error("Unimplemented variadic lowering for CallInst");
+  }
+
+  if (VaList)
+    Builder.CreateLifetimeEnd(VaList, sizeOfAlloca(Ctx, DL, VaList));
+
+  Builder.CreateLifetimeEnd(Alloced, sizeOfAlloca(Ctx, DL, Alloced));
+
+  NewCB->setAttributes(PAL);
+  NewCB->takeName(CB);
+  NewCB->setCallingConv(CB->getCallingConv());
+
+  NewCB->setDebugLoc(DebugLoc());
+
+  // I think this is upsetting the debug handling (DISubprogram attached to more
+  // than one function) Need to move metadata, not copy it?
+  NewCB->copyMetadata(*CB, {LLVMContext::MD_prof, LLVMContext::MD_dbg});
+
+  CB->replaceAllUsesWith(NewCB);
+  CB->eraseFromParent();
+  return Changed;
+}
+
+bool ExpandVariadics::expandVAIntrinsicCall(IRBuilder<> &Builder,
+                                            const DataLayout &DL,
+                                            VAStartInst *Inst) {
+  Function *ContainingFunction = Inst->getFunction();
+  if (ContainingFunction->isVarArg())
+    return false;
+
+  // The last argument is a vaListParameterType
+  Argument *PassedVaList =
+      ContainingFunction->getArg(ContainingFunction->arg_size() - 1);
+
+  // va_start takes a pointer to a va_list, e.g. one on the stack
+  Value *VaStartArg = Inst->getArgList();
+
+  Builder.SetInsertPoint(Inst);
+  if (ABI.VAList->passedInSSARegister()) {
+    Builder.CreateStore(PassedVaList, VaStartArg);
+  } else {
+    // src and dst are both pointers
+    memcpyVAListPointers(DL, Builder, VaStartArg, PassedVaList);
+  }
+
+  Inst->eraseFromParent();
+  return true;
+}
+
+bool ExpandVariadics::expandVAIntrinsicCall(IRBuilder<> &, const DataLayout &,
+                                            VAEndInst *Inst) {
+  // A no-op on all the architectures implemented so far
+  Inst->eraseFromParent();
+  return true;
+}
+
+bool ExpandVariadics::expandVAIntrinsicCall(IRBuilder<> &Builder,
+                                            const DataLayout &DL,
+                                            VACopyInst *Inst) {
+  Builder.SetInsertPoint(Inst);
+  memcpyVAListPointers(DL, Builder, Inst->getDest(), Inst->getSrc());
+  Inst->eraseFromParent();
+  return true;
+}
+
+} // namespace
+
+char ExpandVariadics::ID = 0;
+
+INITIALIZE_PASS(ExpandVariadics, DEBUG_TYPE, "Expand variadic functions", false,
+                false)
+
+ModulePass *llvm::createExpandVariadicsPass(ExpandVariadicsMode Mode) {
+  return new ExpandVariadics(Mode);
+}
+
+PreservedAnalyses ExpandVariadicsPass::run(Module &M, ModuleAnalysisManager &) {
+  return ExpandVariadics(ConstructedMode).runOnModule(M)
+             ? PreservedAnalyses::none()
+             : PreservedAnalyses::all();
+}
+
+ExpandVariadicsPass::ExpandVariadicsPass(OptimizationLevel Level)
+    : ExpandVariadicsPass(Level == OptimizationLevel::O0
```

Comment:
I don't think this works, there's a cyclic dependency here on `LLVMPasses` which defines this. The standard way to handle this should be when we register the pass in the `PassBuilderPipeline`.